### 1. Objective of Notebook 3

This notebook focuses on turning raw financial statements into insights by:

* Computing key financial ratios

* Analyzing multi-year trends

* Answering questions on:

   * Profitability

   * Financial stability

   * Risk

   * Valuation

This notebook directly answers the core problems defined in Notebook 1, using the cleaned data from Notebook 2.

### 2. Load Cleaned Financial Data

In [27]:
import pandas as pd

income = pd.read_csv("apple_income_statement.csv", index_col=0, parse_dates=True)
balance = pd.read_csv("apple_balance_sheet.csv", index_col=0, parse_dates=True)
cashflow = pd.read_csv("apple_cash_flow.csv", index_col=0, parse_dates=True)

income.head()
balance.head()
cashflow.head()



Free Cash Flow  Repurchase Of Capital Stock  Repayment Of Debt  \
2025-09-30    9.876700e+10                -9.071100e+10      -1.093200e+10   
2024-09-30    1.088070e+11                -9.494900e+10      -9.958000e+09   
2023-09-30    9.958400e+10                -7.755000e+10      -1.115100e+10   
2022-09-30    1.114430e+11                -8.940200e+10      -9.543000e+09   
2021-09-30             NaN                          NaN                NaN   

            Issuance Of Debt  Issuance Of Capital Stock  Capital Expenditure  \
2025-09-30      4.481000e+09                        NaN        -1.271500e+10   
2024-09-30      0.000000e+00                        NaN        -9.447000e+09   
2023-09-30      5.228000e+09                        NaN        -1.095900e+10   
2022-09-30      5.465000e+09                        NaN        -1.070800e+10   
2021-09-30               NaN               1.105000e+09                  NaN   

            Interest Paid Supplemental Data  \
2025-09-30                              NaN   
2024-09-30                              NaN   
2023-09-30                     3.803000e+09   
2022-09-30                     2.865000e+09   
2021-09-30                     2.687000e+09   

            Income Tax Paid Supplemental Data  End Cash Position  \
2025-09-30                       4.336900e+10       3.593400e+10   
2024-09-30                       2.610200e+10       2.994300e+10   
2023-09-30                       1.867900e+10       3.073700e+10   
2022-09-30                       1.957300e+10       2.497700e+10   
2021-09-30                                NaN                NaN   

            Beginning Cash Position  ...  Change In Receivables  \
2025-09-30             2.994300e+10  ...          -7.029000e+09   
2024-09-30             3.073700e+10  ...          -5.144000e+09   
2023-09-30             2.497700e+10  ...          -4.170000e+08   
2022-09-30             3.592900e+10  ...          -9.343000e+09   
2021-09-30                      NaN  ...                    NaN   

            Changes In Account Receivables  Other Non Cash Items  \
2025-09-30                   -6.682000e+09         -8.900000e+07   
2024-09-30                   -3.788000e+09         -2.266000e+09   
2023-09-30                   -1.688000e+09         -2.227000e+09   
2022-09-30                   -1.823000e+09          1.006000e+09   
2021-09-30                             NaN                   NaN   

            Stock Based Compensation  Deferred Tax  Deferred Income Tax  \
2025-09-30              1.286300e+10           NaN                  NaN   
2024-09-30              1.168800e+10           NaN                  NaN   
2023-09-30              1.083300e+10           NaN                  NaN   
2022-09-30              9.038000e+09  8.950000e+08         8.950000e+08   
2021-09-30                       NaN -4.774000e+09        -4.774000e+09   

            Depreciation Amortization Depletion  \
2025-09-30                         1.169800e+10   
2024-09-30                         1.144500e+10   
2023-09-30                         1.151900e+10   
2022-09-30                         1.110400e+10   
2021-09-30                                  NaN   

            Depreciation And Amortization  \
2025-09-30                   1.169800e+10   
2024-09-30                   1.144500e+10   
2023-09-30                   1.151900e+10   
2022-09-30                   1.110400e+10   
2021-09-30                            NaN   

            Net Income From Continuing Operations  Company  
2025-09-30                           1.120100e+11    Apple  
2024-09-30                           9.373600e+10    Apple  
2023-09-30                           9.699500e+10    Apple  
2022-09-30                           9.980300e+10    Apple  
2021-09-30                                    NaN    Apple  

[5 rows x 54 columns]

In [ ]:
# Sort by year for proper trend analysis
income = income.sort_index()
balance = balance.sort_index()
cashflow = cashflow.sort_index()


### 3. Profitability Analysis
#### Key Questions Answered

* Is Apple becoming more profitable?

* Are margins improving or declining?

In [18]:
# ratios computed
profitability = pd.DataFrame(index=income.index)

profitability["Gross Margin"] = income["Gross Profit"] / income["Total Revenue"]
profitability["Operating Margin"] = income["Operating Income"] / income["Total Revenue"]
profitability["Net Margin"] = income["Net Income"] / income["Total Revenue"]

profitability



Gross Margin  Operating Margin  Net Margin
2025-09-30      0.469052          0.319708    0.269151
2024-09-30      0.462063          0.315102    0.239713
2023-09-30      0.441311          0.298214    0.253062
2022-09-30      0.433096          0.302887    0.253096
2021-09-30           NaN               NaN         NaN

Insights:

* Gross margin is steadily increasing → the company is earning more from each sale.

* Operating margin is stable and slightly improving → costs are well controlled.

* Net margin improves in 2025 after a small dip → overall profitability is strengthening.

Overall: profitability is solid and trending upward, especially in recent years.

### 4. Growth Trend Analysis

In [20]:
# revenue & profit growth
growth = pd.DataFrame(index=income.index)

growth["Revenue Growth %"] = income["Total Revenue"].pct_change() * 100
growth["Net Income Growth %"] = income["Net Income"].pct_change() * 100

print(growth)


            Revenue Growth %  Net Income Growth %
2025-09-30               NaN                  NaN
2024-09-30         -6.037567           -16.314615
2023-09-30         -1.981920             3.476786
2022-09-30          2.881146             2.894995
2021-09-30          0.000000             0.000000


C:\Users\ACER\AppData\Local\Temp\ipykernel_4988\518622293.py:4: FutureWarning: The default fill_method='pad' in Series.pct_change is deprecated and will be removed in a future version. Either fill in any non-leading NA values prior to calling pct_change or specify 'fill_method=None' to not fill NA values.
  growth["Revenue Growth %"] = income["Total Revenue"].pct_change() * 100
C:\Users\ACER\AppData\Local\Temp\ipykernel_4988\518622293.py:5: FutureWarning: The default fill_method='pad' in Series.pct_change is deprecated and will be removed in a future version. Either fill in any non-leading NA values prior to calling pct_change or specify 'fill_method=None' to not fill NA values.
  growth["Net Income Growth %"] = income["Net Income"].pct_change() * 100


Insights:

* 2023: Both revenue and net income declined → a short-term slowdown.

* 2024: Revenue returned to growth, but net income still fell → rising costs or margin pressure.

* 2025: Strong rebound → revenue grew ~6.4% and net income surged ~19.5%, showing improved efficiency and profitability.

Overall: growth momentum strengthened in 2025, with profits growing much faster than sales—a positive sign.

### 5. Financial Stability & Liquidity Analysis
Key Questions Answered

* Can Apple meet short-term obligations?

* How strong is its balance sheet?

In [22]:
# ratios computed
stability = pd.DataFrame(index=balance.index)

# Current Ratio: Cash / Current Liabilities (simplified)
stability["Current_Ratio"] = (
    balance["Cash And Cash Equivalents"] / balance["Current Liabilities"]
)

# Debt to Assets
stability["Debt_to_Assets"] = balance["Total Liabilities Net Minority Interest"] / balance["Total Assets"]

stability



Current_Ratio  Debt_to_Assets
2025-09-30       0.216952        0.794753
2024-09-30       0.169753        0.843964
2023-09-30       0.206217        0.823741
2022-09-30       0.153563        0.856354
2021-09-30            NaN             NaN

Insights:

Current Ratio: Very low throughout (0.15–0.22), indicating the company may struggle to cover short-term liabilities with current assets. Slight improvement in 2025, but still risky.

Debt-to-Assets: High but declining from 85.6% (2022) to 79.5% (2025), showing the company is gradually reducing reliance on debt.

Overall: The company is highly leveraged and has weak short-term liquidity, though debt reduction is a positive trend.

#### 6. Risk & Leverage Analysis

In [23]:
risk = pd.DataFrame(index=balance.index)
risk["Debt_to_Equity"] = balance["Total Liabilities Net Minority Interest"] / balance["Stockholders Equity"]

risk


Debt_to_Equity
2025-09-30        3.872187
2024-09-30        5.408780
2023-09-30        4.673462
2022-09-30        5.961537
2021-09-30             NaN

Insights:

* Debt-to-Equity: Very high throughout (3.8), meaning the company relies heavily on debt compared to equity.

* Trend: Gradually decreasing from 5.96 (2022) to 3.87 (2025), showing the company is slowly deleveraging.

Overall: Still highly leveraged, but the declining trend is a positive sign for financial stability.

### 7. Cash Flow Strength Analysis
Key Questions Answered

* Does Apple generate real cash?

* Is profit backed by cash flow?

In [11]:
cashflow.rename(columns={
    "Total Cash From Operating Activities": "Operating Cash Flow",
    "Capital Expenditures": "CapEx"
}, inplace=True)

cash_analysis = pd.DataFrame(index=cashflow.index)
cash_analysis["Operating Cash Flow"] = cashflow["Operating Cash Flow"]
cash_analysis["Free_Cash_Flow"] = cashflow["Operating Cash Flow"] - cashflow["Capital Expenditure"]

cash_analysis


Operating Cash Flow  Free_Cash_Flow
2025-09-30         1.114820e+11    1.241970e+11
2024-09-30         1.182540e+11    1.277010e+11
2023-09-30         1.105430e+11    1.215020e+11
2022-09-30         1.221510e+11    1.328590e+11
2021-09-30                  NaN             NaN

Insights:

* Operating Cash Flow (OCF): Fairly stable around $110–122B, showing consistent cash generation from core operations. Slight dip in 2023–2025 but still strong.

* Free Cash Flow (FCF): Also stable around ~121–133B, indicating the company has ample cash after capital expenditures to reinvest, pay debt, or return to shareholders.

Overall: The company maintains strong and consistent cash generation, supporting operations and growth despite minor fluctuations.

### 8. Efficiency & Return Metrics
#### Key Questions Answered

* How efficiently does Apple use its assets and equity?

In [25]:
returns = pd.DataFrame(index=income.index)

returns["ROA"] = income["Net Income"] / balance["Total Assets"]
returns["ROE"] = income["Net Income"] / balance["Stockholders Equity"]

returns


ROA       ROE
2025-09-30  0.311796  1.519130
2024-09-30  0.256825  1.645935
2023-09-30  0.275098  1.560760
2022-09-30  0.282924  1.969589
2021-09-30       NaN       NaN

Insights:

ROA: Strong and fairly stable (26%–31%), showing efficient use of assets to generate profit.

ROE: Extremely high (152%–197%), driven largely by heavy leverage rather than equity growth.

Overall: The company is operationally efficient, but the very high ROE reflects high debt levels, which increase financial risk.

### 9. Valuation Perspective (Fundamentals-Based)
Key Questions Answered

* Does Apple’s performance justify its valuation?

In [16]:
valuation = pd.DataFrame(index=income.index)

valuation["Earnings per Share (EPS)"] = income["Net Income"] / income["Diluted Average Shares"]

valuation



Earnings per Share (EPS)
2025-09-30                  7.464996
2024-09-30                  6.083555
2023-09-30                  6.134053
2022-09-30                  6.113200
2021-09-30                       NaN

Insights:

EPS: Largely flat around 6.1 from 2022–2024, indicating stable earnings per share.

2025: Sharp increase to 7.46, showing strong earnings growth and improved profitability.

Overall: Earnings were stable for several years, then saw a clear acceleration in 2025—a positive signal for shareholders.

### 10. Consolidated Financial Ratios Table

In [26]:
financial_summary = pd.concat(
    [profitability, growth, stability, risk, returns],
    axis=1
)

financial_summary


Gross Margin  Operating Margin  Net Margin  Revenue Growth %  \
2025-09-30      0.469052          0.319708    0.269151               NaN   
2024-09-30      0.462063          0.315102    0.239713         -6.037567   
2023-09-30      0.441311          0.298214    0.253062         -1.981920   
2022-09-30      0.433096          0.302887    0.253096          2.881146   
2021-09-30           NaN               NaN         NaN          0.000000   

            Net Income Growth %  Current_Ratio  Debt_to_Assets  \
2025-09-30                  NaN       0.216952        0.794753   
2024-09-30           -16.314615       0.169753        0.843964   
2023-09-30             3.476786       0.206217        0.823741   
2022-09-30             2.894995       0.153563        0.856354   
2021-09-30             0.000000            NaN             NaN   

            Debt_to_Equity       ROA       ROE  
2025-09-30        3.872187  0.311796  1.519130  
2024-09-30        5.408780  0.256825  1.645935  
2023-09-30        4.673462  0.275098  1.560760  
2022-09-30        5.961537  0.282924  1.969589  
2021-09-30             NaN       NaN       NaN

Insights:

* Profitability: Margins (gross, operating, net) show a clear upward trend, with 2025 the strongest year, indicating improved efficiency and pricing power.

* Growth: Revenue and net income were volatile, declining in 2023–2024, but profitability still improved; 2025 growth data is missing but margins suggest a rebound.

* Liquidity: Current ratio remains very low, signaling weak short-term liquidity throughout the period.

* Leverage: Debt levels are high, but both debt-to-assets and debt-to-equity are declining, showing gradual deleveraging.

* Returns: ROA is strong, while ROE is extremely high, mainly driven by heavy leverage rather than equity growth.

Overall: The company is highly profitable and improving operationally, but still carries liquidity risk and elevated financial leverage.